elip-deterministic-blinding-key.mediawiki

```diff
@@ -0,0 +1,133 @@
+This however increases the amount of data that the user needs to back up.
+
+Otherwise they can be deterministically derived from a BIP32 seed following [https://github.com/satoshilabs/slips/blob/master/slip-0077.md SLIP-77].
+This works for singlesig wallets,
+since the only backup that the user needs is the BIP32 seed (or BIP39 mnemonic).
+However this does not apply to multisig wallets
+which usually consist of multiple xpubs,
+and thus multiple BIP32 seeds from which a descriptor blinding key could be derived.
+SLIP77 also does not work well when there are multiple BIP44 accounts,
+as they will have the same master blinding key.
+
+This document defines a standard way to deterministically derive a descriptor blinding key from an ordinary descriptor.
+
+Deriving the descriptor blinding key from the ordinary descriptor means that multisig participants can avoid needing a key construction and sharing protocol.
+
+==Design==
+
+===Overview===
+We propose a new possible value for the descriptor blinding key
+
+<pre>
+ct(elipXXXX,<DESCRIPTOR>)
+</pre>
+
+Which indicates that descriptor blinding key to be used is a '''view key''' deterministically derived from the ordinary descriptor <code>DESCRIPTOR</code>
+
+===Drawbacks===
+If a wallet uses this standard to derive its descriptor blinding key, anyone knowing the ordinary descriptor will be able to unblind all the corresponding outputs.
+
+Potential adopters of this standard should consider whether this behavior is acceptable.
+
+===Specification===
+The following descriptor:
+
+<pre>
+ct(elipXXXX,<DESCRIPTOR>)
+</pre>
+
+Is equivalent to:
+
+<pre>
+ct(<KEY>,<DESCRIPTOR>)
+</pre>
+
+Where <code>KEY</code> is a ELIP-150 view blinding key derived as follows:
+* If the top-level script expression is <code>combo</code>, fail.
+* Future extensions that support multiple descriptors other than multi-path are not supported.
+* If <code>DESCRIPTOR</code> is multi-path, expand to every single descriptor.
+* For each single descriptor:
+** If it has wildcards, get the definite descriptor at index 2<sup>31</sup>-1<ref>
+We want to use a <code>scriptPubKey</code> that will not be used in a transaction.
+Otherwise it can be used directly by anyone to derive the descriptor blinding key.
+Wallets use <code>scriptPubKey</code>s starting from index 0,
+thus we choose the last non-hardened index,
+which in practice will never be used in a transaction.
+</ref>
+** Derive the <code>scriptPubKey</code> from the definite descriptor<ref>
+Here we use the <code>scriptPubKey</code> because descriptors have fields that might change without affecting the spending conditions.
+For instance changing the fingerprint of one xpub yields the same <code>scriptPubKey</code>s.
+Using the <code>scriptPubKey</code> allows us to treat such descriptors as equivalent.
+</ref>
+* Concatenate all <code>scriptPubKey</code>s obtained before, separating them with <code>OP_INVALIDOPCODE</code>.
```

Comment:
In c52c80d9fd82f9f1471c77255c443d7096fa4e6b:

I would suggest "prefixed with `OP_INVALIDOPCODE`" rather than "separated by". That way even if there is only one scriptpubkey, an invalid opcode will still appear.

Also, is `INVALIDOPCODE` a real opcode? If you would like to pick a specific value i'd choose `OP_RESERVED` or `OP_RETURN`.

Reply:
`OP_INVALIDOPCODE` (0xff) is always invalid if used in scripts, from cores `src/test/script_tests.cpp`:

```
    script = ScriptFromHex("ff88ac"); // Script with OP_INVALIDOPCODE explicit
    BOOST_CHECK(!script.HasValidOps());
```

I'd avoid `OP_RETURN` but `OP_RESERVED`  would also work.

Reply:
Changed